### ISSUE TITLE
web2.auth.wrapper.UnauthenticatedResource does not reply with allowed methods for wrapped resource

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#1827]
|<img alt="troy@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">| troy@... reported|
|-|-|
|Trac ID|trac#1827|
|Type|defect|
|Created|2006-06-15 20:31:14Z|

In web2.auth.wrapper, the UnauthenticatedResource class will accept with only the stock HTTP allowed methods (GET, HEAD, OPTIONS, and TRACE), even if the resource being wrapped may support more.  This is particularly a problem for placing DAV resources inside an HTTPAuthResource wrapper; the first request most DAV clients will send to a DAV server uses the PROPFIND method, and UnauthenticatedResource will respond to that with an HTTP status 405 (Method not allowed) rather than an request to authenticate.  This will cause many DAV clients (at least cadaver, Nautilus, and Konqueror) either to give up entirely or to behave erratically.

One way to fix this would be to have the init function for   UnauthenticatedResource take a second argument, which is the list of allowed methods from the wrapped resource.  That can then be appended the response header.

Patch and example DAV service with this problem attached.

(HTTPAuthResource.locateChild() also appears to have an off-by-one error in it; this patch corrects that as well.)

Attachments:

* [DAVservice.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/53a/53a6593ea063a3cda52a0e5ef1ec89326dd75176/a19ef42bb28c34abccd1afd9bc3126379e787f7d.py) (2459 bytes) - added by troy@... on 2006-06-15 20:32:10Z - DAV service example
* [authwrapper.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/53a/53a6593ea063a3cda52a0e5ef1ec89326dd75176/96928f004335d6a93babeaf3cecc18ad3452c686.patch) (2596 bytes) - added by troy@... on 2006-06-15 20:33:05Z - patch for web2.auth.wrapper
* [authtest.py](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/53a/53a6593ea063a3cda52a0e5ef1ec89326dd75176/20e55a4e75212d2aaa6e6ce6597c862483cb6a0c.py) (3181 bytes) - added by rfmohr@... on 2006-10-12 19:12:15Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__1827 1827
type__defect defect
reporter__troy____ troy@...
priority__normal normal
milestone__ 
branch__ 
branch_author__ 
status__closed closed
resolution__fixed fixed
component__web2 web2
keywords__ 
time__1150403474000000 1150403474000000
changetime__1171149097000000 1171149097000000
version__None None
owner__ 
cc__rfmohr@...
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid set owner to @dreid|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)|@dreid commented|
|-|-|

I think HTTPAuthWrapperResource.allowedMethods should just be a property that proxies to the wrapped resource.  And then HTTPAuthWrapperResource should just raise an UnauthorizedResponse such as the one in [#1814](https://github.com/twisted/twisted/issues/1814).

**Comment by twisted-trac** [comment]
|<img alt="rfmohr@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_3"></a>|rfmohr@... commented|
|-|-|

Now that there is support for the UnauthorizedResponse class, I think we can avoid having to deal with the allowedMethods stuff altogether.

My tests indicate the problem has to do with fact that there is no renderHTTP() method in the UnauthorizedResource class.  So when this resource is rendered, the renderHTTP() method from the RenderMixin class is being called.  This method then looks for a http_PROPFIND method in the UnauthorizedResource class, fails to find it, and finally invokes the allowedMethods() method.

Renaming the render(self, req) method to renderHTTP(self, req) solves this problem.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_4"></a>|@dreid commented|
|-|-|

This is actually solved by having the UnauthorizedResponse correct?  Could you write a test case that demonstrates that it works?

**Comment by twisted-trac** [comment]
|<img alt="rfmohr@...'s avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_5"></a>|rfmohr@... commented|
|-|-|

I have attached a script that will illustrate the problem.  If the render() method is renamed to renderHTTP() in the UnauthorizedResource class, the script will output the correct results.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_6"></a>|@dreid commented|
|-|-|

I think a better solution is to have the wrapper resource raise or return an UnauthorisedResponse directly instead of using UnauthorizedResource.

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_7"></a>|@dreid commented|
|-|-|

addressed in [branches/web2-auth-request-2042](branches/web2-auth-request-2042)

**Comment by twisted-trac** [comment]
|[<img alt="dreid's avatar" src="https://avatars.githubusercontent.com/u/48695?s=50" width="50" height="50">](https://github.com/dreid)<a name="note_8"></a>|@dreid set status to `closed`|
|-|-|

(In [19618]) Merge web2-auth-request-2042

Author: dreid
Reviewer: therve

Fixes [#2042](https://github.com/twisted/twisted/issues/2042)
Fixes [#1827](https://github.com/twisted/twisted/issues/1827)

Address deficiencies in the web2 HTTPAuthWrapper implementation.  
Add two new interfaces IAuthenticatedRequest, and IHTTPUser that
may be used to access the avatar for the authenticated user and 
get information about that user from the avatar.  

Also address two bugs in HTTPAuthWrapper that caused incorrect
reporting of allowedMethods and the first segment of the url to
be lost when returning the wrapped resource from locateChild.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
